neps/nep-0508.md

```diff
@@ -0,0 +1,231 @@
+
+## Future possibilities
+
+### Dynamic resharding
+
+As noted above, dynamic resharding is out of scope for this NEP and should be implemented in the future. Dynamic resharding includes the following but not limited to:
+
+* Automatic determination of split boundary based on parameters like traffic, gas usage, state size, etc.
+* Automatic shard splitting and merging
+
+### Localization of resharding event to specific shard
+
+As of today, at the RocksDB storage layer, we have the ShardUId, i.e. the ShardId along with the ShardVersion, as a prefix in the key of trie state and flat state. During a resharding event, we increment the ShardVersion by one, and effectively remap all the current parent shards to new child shards. This implies we can't use the same underlying key value pairs for store and instead would need to duplicate the values with the new ShardUId prefix, even if a shard is unaffected and not split.
+
+In the future, we would like to potentially change the schema in a way such that only the shard that is splitting is impacted by a resharding event, so as to avoid additonal work done by nodes tracking other shards.
+
+### Other useful features
+
+* Account colocation for low latency across account call
+* Removal of shard uids and introducing globally unique shard ids
+* Shard on demand
+
+## Consequences
+
+### Positive
+
+* Workload across shards will be more evenly distributed.
+* Required space to maintain state (either in memory or in persistent disk) will be smaller. This is useful for in-memory tries.
+* State sync overhead will be smaller with smaller state size.
+
+### Neutral
+
+* Number of shards would increase.
+* Underlying trie structure and data structure are not going to change.
+* Resharding will create dependency on flat state snapshots.
+
+### Negative
+
+* The resharding process, as of now, is not fully automated. Analyzing shard data, determining the split boundary, and triggering an actual shard split all need to be manually curated and tracked.
+* During resharding, a node is expected to require more resources as it will first need to copy state data from the parent shard to the child shard, and then will have to apply trie and flat state changes twice, once for the parent shard and once for the child shards.
+* Increased potential for apps and tools to break without proper shard layout change handling.
+
+### Backwards Compatibility
+
+We do not expect anything to break with this change. Yet, shard splitting can introduce additional complexity on replayability. For instance, as target shard of a receipt and belonging shard of an account can change with shard splitting, shard splitting must be replayed along with transactions at the exact epoch boundary.
```

Comment:
I would add that all tools assuming `num_shards = 4` will break, if tool maintainters monitor NEPs

Reply:
Also, light clients cannot get the shard boundary from the chain itself. The chain itself unfortunately does not encode the shard layout. It only encodes the protocol version, and then we need to figure out the shard layout ourselves. So light clients would have to be updated.